src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetHttpModule.cs

```diff
@@ -0,0 +1,151 @@
+    {
+        private static readonly ILog Log = LogProvider.GetLogger(typeof(AspNetHttpModule));
+
+        private readonly string _httpContextKey;
+        private readonly string _operationName;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AspNetHttpModule"/> class.
+        /// </summary>
+        /// <param name="operationName">The operation name to be used for the trace/span data generated</param>
+        public AspNetHttpModule(string operationName)
+        {
+            _operationName = operationName ?? throw new ArgumentNullException(nameof(operationName));
+
+            _httpContextKey = string.Concat("__Datadog.Trace.ClrProfiler.Integrations.AspNetHttpModule-", _operationName);
+        }
+
+        /// <inheritdoc />
+        public void Init(HttpApplication context)
+        {
+            if (!Instrumentation.ProfilerAttached)
+            {
+#if DEBUG
+                context.Context.AddError(new ApplicationException("Datadog Profiler not attached to current process."));
+#endif
+
+                Log.Warn("Datadog ASP.NET HttpModule Profiler not attached");
+
+                return;
+            }
+
+            context.BeginRequest += OnBeginRequest;
+            context.EndRequest += OnEndRequest;
+            context.Error += OnError;
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            // Nothing to do...
+        }
+
+        private void OnBeginRequest(object sender, EventArgs eventArgs)
+        {
+            Scope scope = null;
+
+            try
+            {
+                if (!TryGetContext(sender, out var httpContext))
+                {
+                    return;
+                }
+
+                scope = Tracer.Instance.StartActive(_operationName);
+
+                var contextAdapter = HttpContextTagAdapter.Create(httpContext);
+
+                var decorator = DefaultSpanDecorationBuilder.Create()
+                                                            .With(contextAdapter.AllWebSpanDecorator())
+                                                            .With(contextAdapter.ResourceNameDecorator())
+                                                            .Build();
+
+                scope.Span.DecorateWith(decorator);
+
+                httpContext.Items[_httpContextKey] = scope;
+            }
+            catch (Exception ex)
+            {
+                // Dispose here, as the scope won't be in context items and won't get disposed on request end in that case...
+                scope?.TryDispose();
+
+                Log.ErrorException("Datadog ASP.NET HttpModule instrumentation error", ex);
+            }
+        }
+
+        private void OnEndRequest(object sender, EventArgs eventArgs)
+        {
+            try
+            {
+                if (!TryGetContext(sender, out var httpContext) ||
+                    !httpContext.Items.TryGetValueOrDefaultAs<Scope>(_httpContextKey, out var scope))
```

Comment:
Only if the concrete implementation is a HashTable or similar...those aren't the only IDictionary implementations

Reply:
Any implementation of [`IDictionary.Item[object]`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.idictionary.item) they throws instead of returning null on missing key is broken.